### PR TITLE
Avoid repeated precompilation when loading from non-relocatable cachefiles

### LINKED ISSUE #53859: Parallel precompile now seems to fail precompiling the package itself
```
(Foo) pkg> precompile
   Installed Example ─ v0.5.3
Precompiling all packages...
  1 dependency successfully precompiled in 1 seconds

(Foo) pkg> st
Status `~/julia/Foo/Project.toml`
  [7876af07] Example v0.5.3

julia> using Foo # should have precompiled `Foo` already
[ Info: Precompiling Foo [16a8b8d3-1370-5883-bb6f-f9266484db3f] 
```

### PR DESCRIPTION
**Pull request by fatteneder**
Fixes https://github.com/JuliaLang/julia/issues/53859#issuecomment-2027352004, which was actually fixed before in https://github.com/JuliaLang/julia/pull/52346, but https://github.com/JuliaLang/julia/pull/52750 undid that fix.

This PR does not directly address #53859, because I can not reproduce it atm.

---

The `@depot` resolution logic for `include()` files is adjusted as follows:

1. (new behavior) If the cache is not relocatable because of an absolute path, we ignore that path for the depot search. Recompilation will be triggered by `stale_cachefile()` if that absolute path does not exist. Previously this caused any `@depot` tags to be not resolved and so trigger recompilation.
2. (new behavior) If we can't find a depot for a relocatable path, we still replace it with the depot we found from other files. Recompilation will be triggered by `stale_cachefile()` because the resolved path does not exist.
3. (this behavior is kept) We require that relocatable paths all resolve to the same depot.
4. (new behavior) We no longer use the first matching depot for replacement, but instead we explicitly check that all resolve to the same depot. This has two reasons:
  - We want to scan all source files anyways in order to provide logs for 1. and 2. above, so the check is free.
  - It is possible that a depot might be missing source files. Assume that we have two depots on `DEPOT_PATH`, `depot_complete` and `depot_incomplete`.
    If `DEPOT_PATH=["depot_complete","depot_incomplete"]` then no recompilation shall happen, because `depot_complete` will be picked.
    If `DEPOT_PATH=["depot_incomplete","depot_complete"]` we trigger recompilation and hopefully a meaningful error about missing files is thrown. If we were to just select the first depot we find, then whether recompilation happens would depend on whether the first relocatable file resolves to `depot_complete` or `depot_incomplete`.

### CODE REVIEW

File: base/loading.jl [hunk 1]
```diff
@@ -3150,31 +3150,70 @@ function parse_cache_header(f::IO, cachefile::AbstractString)
...
+    if multiple_depots_found
+        @debug("Some include() files from $cachefile are distributed over multiple depots", srcfiles, _group=:relocatable)
+    elseif !isnothing(srcdepot)
         for inc in includes_srcfiles
-            inc.filename = restore_depot_path(inc.filename, srcfiles_depot)
+            inc.filename = restore_depot_path(inc.filename, srcdepot)
         end
     end
+
+    # each include_dependency() file is allowed to relocate to a separate depot, #52161
```
Comment (staticfloat) [inline base/loading.jl:3210]: I think it is worth pointing out here that this is in contrast to `include()` files, where we do require they all come from the same depot: https://github.com/JuliaLang/julia/pull/53905/files#diff-ec80afe9f8ff182765c73ed3fc35a103b561f873a52b1dfe1ce135e4e5f9a2cfR3169 as the reader may not immediately understand that `include()` and `include_dependency()` are different in this way.

**Review by staticfloat — APPROVED** [review]

File: test/relocatedepot.jl [hunk 2]
```diff
@@ -123,6 +125,25 @@ if !test_relocated_depot
...
     end
 
+    @testset "precompile RelocationTestPkg4" begin
+        # test for #52346 and https://github.com/JuliaLang/julia/issues/53859#issuecomment-2027352004
+        # If a pkgimage is not relocatable, no repeated precompilation should occur.
+        pkgname = "RelocationTestPkg4"
+        test_harness(empty_depot_path=false) do
+            push!(LOAD_PATH, @__DIR__)
+            # skip this dir to make the pkgimage not relocatable
+            filter!(!=(@__DIR__), DEPOT_PATH)
```
Comment (fatteneder) [inline test/relocatedepot.jl:135]:
```suggestion
            filter!(!=(@__DIR__), DEPOT_PATH)
            @show LOAD_PATH
            @show DEPOT_PATH
```

File: test/relocatedepot.jl [hunk 3]
```diff
@@ -123,6 +125,27 @@ if !test_relocated_depot
...
```
Comment (fatteneder) [inline test/relocatedepot.jl:135]:
```suggestion
            filter!(DEPOT_PATH) do depot
                !startswith(@__DIR__, depot)
            end
```

File: base/loading.jl [hunk 4]
```diff
@@ -1757,11 +1757,15 @@ associated cache is relocatable.
 """
 function isrelocatable(pkg::PkgId)
     path = compilecache_path(pkg)
+    @show pkg
```
Comment (giordano) [inline base/loading.jl:1760]: This PR adds a bunch of `@show`s, also below and in other files, presumably they should be removed?

Comment (fatteneder) [reply base/loading.jl:1760]: Yes. These are leftovers from debugging CI, I just did not have time yet to rebase.

**Pull request merged**